packages/oid4vci-rest-client/src/agent/OID4VCIRestClient.ts

```diff
@@ -0,0 +1,65 @@
+import { fetch } from 'cross-fetch'
+import {
+  IOID4VCIRestClient,
+  IVCIClientCreateOfferUriRequest,
+  IVCIClientCreateOfferUriRequestArgs,
+  IVCIClientCreateOfferUriResponse,
+} from '../types/IOID4VCIRestClient'
+import Debug from 'debug'
+import { IAgentPlugin } from '@veramo/core'
+
+const debug = Debug('sphereon:ssi-sdk.oid4vci-rest-client')
+
+/**
+ * @beta
+ */
+export class OID4VCIRestClient implements IAgentPlugin {
+  readonly methods: IOID4VCIRestClient = {
+    vciClientCreateOfferUri: this.vciClientCreateOfferUri.bind(this),
+  }
+
+  private readonly baseUrl?: string
+
+  constructor(args?: { baseUrl?: string; definitionId?: string }) {
+    if (args?.baseUrl) {
+      this.baseUrl = args.baseUrl
+    }
+  }
+
+  private async vciClientCreateOfferUri(args: IVCIClientCreateOfferUriRequestArgs): Promise<IVCIClientCreateOfferUriResponse> {
+    if (!args.credentials) {
+      throw new Error("Can't generate the credential offer uri without credentials param present.")
+    }
+    const baseUrl = this.checkBaseUrlParameter(args.baseUri)
+    const request: IVCIClientCreateOfferUriRequest = {
+      credentials: args.credentials,
+    }
+    if (args.grants) {
+      request['grants'] = args.grants
+    }
+    const url = this.uriWithBase(`webapp/credential-offers`, baseUrl)
+    debug(`OID4VCIRestClient is going to send request: ${JSON.stringify(request)} to ${url}`)
+    const origResponse = await fetch(url, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(request),
+    })
+    return await origResponse.json()
+  }
+
+  private uriWithBase(path: string, baseUrl?: string): string {
+    if (!this.baseUrl && !baseUrl) {
+      throw new Error('You have to provide baseUrl')
+    }
+    return baseUrl ? `${baseUrl}${path.startsWith('/') ? path : '/' + path}` : `${this.baseUrl}${path.startsWith('/') ? path : '/' + path}`
+  }
+
+  private checkBaseUrlParameter(baseUrl?: string): string {
+    if (!baseUrl && !this.baseUrl) {
+      throw new Error('No base url has been provided')
+    }
+    return baseUrl ? baseUrl : (this.baseUrl as string)
```

Comment:
this.baseUrl is a string?

Reply:
An assertBaseUrl method that checks against this and against a param makes sense, as this is a common ocurence in this class it seems

Reply:
this is a string, but since it's optional, I can't pass it without the `as` keyword. anyways, changed this method to just assert that we have a `baseUrl` else throws an error